yt/frontends/swift/io.py

```diff
@@ -0,0 +1,168 @@
+import numpy as np
+
+from yt.frontends.sph.io import \
+    IOHandlerSPH
+
+class IOHandlerSwift(IOHandlerSPH):
+    _dataset_type = "swift"
+
+    def __init__(self, ds, *args, **kwargs):
+        super(IOHandlerSwift, self).__init__(ds, *args, **kwargs)
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
+        raise NotImplementedError
+
+    def _read_particle_coords(self, chunks, ptf):
+        # This will read chunks and yield the results.
+        chunks = list(chunks)
+        data_files = set([])
+        for chunk in chunks:
+            for obj in chunk.objs:
+                data_files.update(obj.data_files)
+        for data_file in sorted(data_files, key=lambda x: x.filename):
+            si, ei = data_file.start, data_file.end
+            f = h5py.File(data_file.filename, "r")
+            # This double-reads
+            for ptype, field_list in sorted(ptf.items()):
+                if data_file.total_particles[ptype] == 0:
+                    continue
+                x = f["/%s/Coordinates" % ptype][si:ei, 0].astype("float64")
+                y = f["/%s/Coordinates" % ptype][si:ei, 1].astype("float64")
+                z = f["/%s/Coordinates" % ptype][si:ei, 2].astype("float64")
+                if ptype == self.ds._sph_ptype:
+                    hsml = f[
+                        "/%s/SmoothingLength" % ptype][si:ei].astype("float64")
+                else:
+                    hsml = 0.0
+                yield ptype, (x, y, z), hsml
+            f.close()
+
+    def _yield_coordinates(self, data_file, needed_ptype=None):
+        si, ei = data_file.start, data_file.end
+        f = h5py.File(data_file.filename)
+        pcount = f["/Header"].attrs["NumPart_ThisFile"][:].astype("int")
+        np.clip(pcount - si, 0, ei - si, out=pcount)
+        pcount = pcount.sum()
+        for key in f.keys():
+            if not key.startswith("PartType"):
+                continue
+            if "Coordinates" not in f[key]:
+                continue
+            if needed_ptype and key != needed_ptype:
+                continue
+            ds = f[key]["Coordinates"][si:ei,...]
+            dt = ds.dtype.newbyteorder("N") # Native
+            pos = np.empty(ds.shape, dtype=dt)
+            pos[:] = ds
+            yield key, pos
+        f.close()
+
+    def _get_smoothing_length(self, data_file, position_dtype, position_shape):
+        ptype = self.ds._sph_ptype
+        ind = int(ptype[-1])
+        si, ei = data_file.start, data_file.end
+        with h5py.File(data_file.filename) as f:
+            pcount = f["/Header"].attrs["NumPart_ThisFile"][ind].astype("int")
+            pcount = np.clip(pcount - si, 0, ei - si)
+            ds = f[ptype]["SmoothingLength"][si:ei,...]
+            dt = ds.dtype.newbyteorder("N") # Native
+            if position_dtype is not None and dt < position_dtype:
+                # Sometimes positions are stored in double precision
```

Comment:
Positions are _always_ doubles in SWIFT. Smoothing lengths are _always_ singles.